File: auth_service/src/middleware/getUserSession.ts
import { Response, NextFunction } from 'express';
import { AuthenticatedRequest, SessionData } from '../types';

/**
 * Middleware that extracts session data from the `session` cookie
 * set by the API gateway proxy and attaches it to `req.user`.
 *
 * The gateway serializes `SessionData` as URI-encoded JSON in a
 * `session` cookie on every proxied request.
 *
 * @param req - Express request with cookies parsed by cookie-parser
 * @param res - Express response (401 if cookie is missing or invalid)
 * @param next - Called with populated `req.user` on success
 */
export async function getUserSession(
  req: AuthenticatedRequest,
  res: Response,
  next: NextFunction
): Promise<void> {
  const sessionCookie = req.cookies?.session;
  if (!sessionCookie) return next();

  try {
    const session: SessionData = JSON.parse(decodeURIComponent(sessionCookie));
    req.user = session;
    next();
  } catch {
    res.status(401).json({ error: 'Invalid session cookie' });
  }
}

File: video_service/app/middleware/get_user_session.py
import json
import logging
from urllib.parse import unquote

from fastapi import Request

from app.models.session import SessionData

logger = logging.getLogger(__name__)


async def get_user_session(request: Request) -> SessionData | None:
    """FastAPI dependency that extracts session data from the ``session`` cookie.

    The API gateway serializes ``SessionData`` as URI-encoded JSON in a
    ``session`` cookie on every proxied request.

    Returns ``None`` if the cookie is missing (allows unauthenticated access).
    Raises no exception on invalid cookie — logs and returns ``None``.
    """
    session_cookie = request.cookies.get("session")
    if not session_cookie:
        return None

    try:
        session = SessionData.model_validate_json(unquote(session_cookie))
        return session
    except Exception:
        logger.warning("Invalid session cookie")
        return None
